### PR TITLE
🌱 feature: enable verification for provenance

### LINKED ISSUE #2747: Feature: add verification step for SLSA 3 provenance with slsa-verifier
**Is your feature request related to a problem? Please describe.**
This is a feature.

**Describe the solution you'd like**
[slsa-verifier](https://github.com/slsa-framework/slsa-verifier) is a tool to verify provenance from SLSA-compliant builders. As the scorecard uses one of the SLSA-compliant builders, I thought that it'd be good to use slsa-verifier in the verification step to verify provenance as we did in the ko project. 

> _https://github.com/ko-build/ko/blob/2860fcf5eec19d9716aaa9c26394da0b1e7b13de/.github/workflows/release.yml#L70_

I'm willing to do this, assign it to me please ✋ 

**Describe alternatives you've considered**

**Additional context**

/cc @asraa @laurentsimon @ianlewis

### PR DESCRIPTION
**Pull request by developer-guy**
#### What kind of change does this PR introduce?

(Is it a bug fix, feature, docs update, something else?)

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

There is no verification step for provenance

#### What is the new behavior (if this is a feature change)?**

Provenance will be verified right after the generated during the build.

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #2747 

#### Special notes for your reviewer

- Successful workflow: https://github.com/developer-guy/scorecard/actions/runs/4457375709/jobs/7828447299

PTAL @laurentsimon 

#### Does this PR introduce a user-facing change?

Nope.

### CODE REVIEW

**Comment by developer-guy** [comment]
Since this workflow is designed to be executed in the workflow_dispatch event, I could not understand how I can use `source-tag` or `source-branch` in the verify-artifact command of slsa-verifier, if you enlighten me on this I can fix this. cc @ianlewis @asraa

**Comment by ianlewis** [comment]
> Since this workflow is designed to be executed in the workflow_dispatch event, I could not understand how I can use `source-tag` or `source-branch` in the verify-artifact command of slsa-verifier, if you enlighten me on this I can fix this. cc @ianlewis @asraa

It looks like the version is determined by the `scripts/version-ldflags` script. You may need to alter this script or write your own in order to get the raw version/tag out.

**Comment by asraa** [comment]
> You may need to alter this script or write your own in order to get the raw version/tag out.

I think actually this one is just used for testing, and not for uploading to the release assets (I think this workflow does that: https://github.com/ossf/scorecard/blob/dfc24396250d6ae74e183b6c8f345ca916d8824b/.github/workflows/goreleaser.yaml#L75). I think therefore that `--source-branch` and `--source-tag` must be kept empty in this workflow.

File: .github/workflows/goreleaser.yaml [hunk 1]
```diff
@@ -76,3 +76,35 @@ jobs:
...
+          name: ${{ needs.provenance.outputs.go-binary-name }}
+
+      - name: Verify assets
+        env:
+          ARTIFACT: ${{ needs.provenance.outputs.go-binary-name }}
+          ATT_FILE_NAME: "${{ needs.provenance.outputs.go-binary-name }}.intoto.jsonl"
+        run: |
+          set -euo pipefail
+              
+          echo "Verifying $ARTIFACT using $ATT_FILE_NAME"
```
Comment (laurentsimon) [inline .github/workflows/goreleaser.yaml:106]: you need to use the list of artifacts from goreleaser and verify them all

Comment (developer-guy) [reply .github/workflows/goreleaser.yaml]: I just updated accordingly, can you please take a look?

File: .github/workflows/slsa-goreleaser.yml [hunk 2]
```diff
@@ -21,15 +22,46 @@ jobs:
...
       go-version: 1.19
       evaluated-envs: "VERSION_LDFLAGS:${{needs.args.outputs.ldflags}}"
+
+  verification:
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    permissions: read-all
+    steps:
+      # Note: this will be replaced with the GHA in the future.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply .github/workflows/slsa-goreleaser.yml]: fixed, thank you.

File: .github/workflows/goreleaser.yaml [hunk 3]
```diff
@@ -76,3 +76,35 @@ jobs:
     with:
       base64-subjects: "${{ needs.goreleaser.outputs.hashes }}"
       upload-assets: true # upload to a new release
+
+  verification:
+    needs: [provenance]
+    runs-on: ubuntu-latest
+    permissions: read-all
+    steps:
+      # Note: this will be replaced with the GHA in the future.
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply .github/workflows/goreleaser.yaml]: fixed, thank you.

File: .github/workflows/goreleaser.yaml [hunk 4]
```diff
@@ -72,7 +72,39 @@ jobs:
...
+        env:
+          CHECKSUMS: ${{ needs.goreleaser.outputs.hashes }}
+          PROVENANCE: "${{ needs.provenance.outputs.provenance-name }}"
+        run: |
+          set -euo pipefail
+          checksums=$(echo "$CHECKSUMS" | base64 -d)
+          while read -r line; do
+              fn=$(echo $line | cut -d ' ' -f2)
+              echo "Verifying $fn"
+              ./slsa-verifier-linux-amd64 -artifact-path "$fn" \
```
Comment (laurentsimon) [inline .github/workflows/goreleaser.yaml:106]: this is using v1.x CLI argument. Can you update to v2.x?

Comment (developer-guy) [reply .github/workflows/goreleaser.yaml:106]: What does it mean ? 🤔

Comment (laurentsimon) [reply .github/workflows/goreleaser.yaml:106]: You're installing with `installer@v2.x1, but the command you're running assumes the old CLI argument from v1.x.
Also, the Action always installs the binary as `slsa-verifier` and it's added to the `PATH`, so you need to run `slsa-verifier`, not `./slsa-verifier-linux-amd64`

Comment (developer-guy) [reply .github/workflows/goreleaser.yaml:106]: gotcha!

**Comment by developer-guy** [comment]
somehow the following line does not work:

```
CHECKSUMS: ${{ needs.goreleaser.outputs.hashes }}
```

I tested this on my fork, got the following error 🤔 PTAL @laurentsimon 

> _https://github.com/developer-guy/scorecard/actions/runs/4477668554/jobs/7870025355_

**Comment by developer-guy** [comment]
> somehow the following line does not work:
> 
> ```
> CHECKSUMS: ${{ needs.goreleaser.outputs.hashes }}
> ```
> 
> I tested this on my fork, got the following error 🤔 PTAL @laurentsimon
> 
> > _[developer-guy/scorecard/actions/runs/4477668554/jobs/7870025355](https://github.com/developer-guy/scorecard/actions/runs/4477668554/jobs/7870025355)_

my bad 🤦 I forgot to add `goreleaser` to the needs list, fixed.

**Review by laurentsimon — APPROVED** [review]

**Review by ianlewis — APPROVED** [review]

File: .github/workflows/slsa-goreleaser.yml [hunk 5]
```diff
@@ -21,15 +22,57 @@ jobs:
...
+        env:
+          ARTIFACT: ${{ needs.build.outputs.go-binary-name }}
+          ATT_FILE_NAME: "${{ needs.build.outputs.go-binary-name }}.intoto.jsonl"
+        run: |
+          set -euo pipefail
+              
+          echo "Verifying $ARTIFACT using $ATT_FILE_NAME"
+          slsa-verifier verify-artifact --provenance-path "$ATT_FILE_NAME" \
+                                        --source-uri "github.com/$GITHUB_REPOSITORY" \
+                                        "$ARTIFACT"
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply .github/workflows/slsa-goreleaser.yml]: thank you 👍

**Comment by developer-guy** [comment]
kindly ping @naveensrinivasan

**Comment by developer-guy** [comment]
kindly ping folx ☝️

**Review by laurentsimon — APPROVED** [review]

**Review by naveensrinivasan — APPROVED** [review]

**Comment by naveensrinivasan** [comment]
@developer-guy  I apologize for the delay. Can you please rebase and fix it?

**Comment by naveensrinivasan** [comment]
> @developer-guy I apologize for the delay. Can you please rebase and fix the conflict?

@developer-guy A friendly ping

**Comment by developer-guy** [comment]
> > @developer-guy I apologize for the delay. Can you please rebase and fix the conflict?
> 
> 
> 
> @developer-guy A friendly ping

Of course give me 5 min 🫡

**Comment by developer-guy** [comment]
all looks good @naveensrinivasan <3

**Review by laurentsimon — APPROVED** [review]

**Pull request merged**